pkg/etcd/stream.go

```diff
@@ -44,25 +44,27 @@ func NewStream(logger kitlog.Logger, client etcdGetter, opt StreamOptions) (<-ch
 		defer wg.Done()
 
 	Watch:
+		for {
+			logger.Log("event", "watch_start")
+			for resp := range client.Watch(clientv3.WithRequireLeader(ctx), "/", clientv3.WithPrefix()) {
+				if resp.Err() != nil {
+					logger.Log("error", resp.Err(), "msg", "received error from etcd watcher")
+				}
 
-		logger.Log("event", "watch_start")
-		for resp := range client.Watch(clientv3.WithRequireLeader(ctx), "/", clientv3.WithPrefix()) {
-			if resp.Err() != nil {
-				logger.Log("error", resp.Err(), "msg", "received error from etcd watcher")
-			}
-
-			for _, event := range resp.Events {
-				if includes(opt.Keys, string(event.Kv.Key)) {
-					out <- event.Kv
+				for _, event := range resp.Events {
+					if includes(opt.Keys, string(event.Kv.Key)) {
+						out <- event.Kv
+					}
 				}
 			}
-		}
 
-		select {
-		case <-ctx.Done():
-			logger.Log("event", "watch_stop", "msg", "context expired, stopping stream")
-		case <-time.After(opt.WatchRetryInterval):
-			goto Watch
+			select {
+			case <-ctx.Done():
+				logger.Log("event", "watch_stop", "msg", "context expired, stopping stream")
+				break Watch
```

Comment:
This is almost identical to what I just tried, but I had a bare `break` here.

Unless I've misunderstood how this is meant to work, I think `break Watch` is going to behave different from what you had before. I thought that if the context is cancelled you want to exit completely rather than attempting to set up the watch again?

Reply:
I think the `break Watch` means break the entire for loop, not just a single iteration. A bare break would cause us to break out of the select and restart the next outer loop iteration, while this will break out of the outer loop itself.

Reply:
Ah right yeah this is me confusing the goto behaviour with the break behaviour. This makes total sense.